File: 프로그래머스/Lv.4/1843. 사칙연산/사칙연산.py
def solution(arr):
    INF = int(1e9)
    n = len(arr) // 2 + 1

    min_dp = [[INF] * n for _ in range(n)]
    max_dp = [[-INF] * n for _ in range(n)]
    
    for x in range(n):
        max_dp[x][x] = int(arr[x * 2])
        min_dp[x][x] = int(arr[x * 2])
    
    for step in range(1, n):
        for i in range(n - step):
            j = i + step
            
            for k in range(i, j):
                if arr[k * 2 + 1] == '+':
                    max_dp[i][j] = max(max_dp[i][j], max_dp[i][k] + max_dp[k + 1][j])
                    min_dp[i][j] = min(min_dp[i][j], min_dp[i][k] + min_dp[k + 1][j])
                else:
                    max_dp[i][j] = max(max_dp[i][j], max_dp[i][k] - min_dp[k + 1][j])
                    min_dp[i][j] = min(min_dp[i][j], min_dp[i][k] - max_dp[k + 1][j])
                
    
    return max_dp[0][n - 1]

File: 프로그래머스/Lv.4/1843. 사칙연산/사칙연산.java
import java.util.Arrays;

class Solution {
    
    private static final int INFINITY = 1_000_000_000;
    
    public int solution(String arr[]) {
        int size = arr.length / 2 + 1;
        
        int[][] maxDp = createDp(size, -INFINITY, arr);
        int[][] minDp = createDp(size, INFINITY, arr);
        
        for (int step = 1; step < size; step++) {
            for (int i = 0; i < size - step; i++) {
                int j = i + step;
                
                for (int k = i; k < j; k++) {
                    if (arr[k * 2 + 1].equals("+")) {
                        maxDp[i][j] = Math.max(maxDp[i][j], maxDp[i][k] + maxDp[k + 1][j]);
                        minDp[i][j] = Math.min(minDp[i][j], minDp[i][k] + minDp[k + 1][j]);
                    } else {
                        maxDp[i][j] = Math.max(maxDp[i][j], maxDp[i][k] - minDp[k + 1][j]);
                        minDp[i][j] = Math.min(minDp[i][j], minDp[i][k] - maxDp[k + 1][j]);
                    }
                }
            }
        }
        
        return maxDp[0][size - 1];
    }
    
    private int[][] createDp(int size, int number, String[] arr) {
        int[][] dp = new int[size][size];
        
        for (int[] row : dp) {
            Arrays.fill(row, number);
        }
        for (int index = 0; index < size; index++) {
            dp[index][index] = Integer.parseInt(arr[index * 2]);
        }
        
        return dp;
    }
}
